File: app/src/main/java/com/mytaxiapp/ui/listvehicles/VehicleListAdapter.java
package com.mytaxiapp.ui.listvehicles;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mytaxiapp.R;
import com.mytaxiapp.data.model.Placemark;
import com.mytaxiapp.ui.map.MapActivity;
import com.mytaxiapp.util.Logger;

import java.util.ArrayList;
import java.util.List;


/**
 * @author Tosin Onikute.
 */

public class VehicleListAdapter
        extends RecyclerView.Adapter<VehicleListAdapter.ViewHolder> {

    private final Logger logger = Logger.getLogger(getClass());
    private final TypedValue mTypedValue = new TypedValue();
    private int mBackground;
    private ArrayList<Placemark> placemarks;
    private Context context;

    private String caddress;
    private String cname;
    private String cinterior;
    private String cfuel;

    private String INTERIOR = "INTERIOR";
    private String VIN = "VIN";
    private String FUEL = "FUEL";
    private String ENGINE = "ENGINE";

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final View mView;
        public final TextView textAddress;
        public final TextView textName;
        public final TextView textInterior;
        public final TextView textFuel;

        public ViewHolder(View view) {
            super(view);
            mView = view;

            textAddress = (TextView) view.findViewById(R.id.address);
            textName = (TextView) view.findViewById(R.id.name);
            textInterior = (TextView) view.findViewById(R.id.interior_vin);
            textFuel = (TextView) view.findViewById(R.id.fuel_engine);
        }

        @Override
        public String toString() {
            return super.toString();
        }
    }

    public String getValueAt(int position) {
        return String.valueOf(placemarks.get(position).getName());
    }

    public VehicleListAdapter(Context context, ArrayList<Placemark> placemarks) {
        context.getTheme().resolveAttribute(R.attr.selectableItemBackground, mTypedValue, true);
        this.context = context;
        this.placemarks = placemarks;
        mBackground = mTypedValue.resourceId;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.vehicle_list, parent, false);
        view.setBackgroundResource(mBackground);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        /* Set your values */
        final Placemark model = (Placemark) placemarks.get(position);

        caddress = "";
        cname = "";
        cinterior = "";
        cfuel = "";

        if (model.getAddress() != null) {
            caddress = model.getAddress();
        }
        if (model.getName() != null) {
            cname = model.getName();
        }
        if (model.getInterior() != null && model.getVin() != null) {
            cinterior = INTERIOR + " : " + model.getInterior() + " | " + VIN + " : " + model.getVin();
        }
        if (model.getFuel() != null && model.getEngineType() != null) {
            cfuel = FUEL + " : " + model.getFuel() + "  |  " + ENGINE + " : " + model.getEngineType();
        }

        holder.textAddress.setText(caddress);
        holder.textName.setText(cname);
        holder.textInterior.setText(cinterior);
        holder.textFuel.setText(cfuel);

        // launch the detail activity to show vehicle information
        holder.mView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Context context = v.getContext();
                Intent intent = new Intent(context, MapActivity.class);
                intent.putExtra("position", holder.getAdapterPosition());
                Activity activity = (Activity) v.getContext();
                activity.startActivityForResult(intent, 500);
                activity.overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left);
            }
        });

    }

    @Override
    public int getItemCount() {
        return (null != placemarks ? placemarks.size() : 0);
    }

    public void addAll(List<Placemark> data){
        notifyDataSetChanged();
    }

    public void add(Placemark data){
        notifyDataSetChanged();
        placemarks.add(data);
    }


    public Placemark getItemPos(int pos){
        return placemarks.get(pos);
    }

    public void clear(){
        placemarks.clear();
    }

}

File: app/src/main/java/com/mytaxiapp/ui/listvehicles/VehicleListAdapter.kt
package com.mytaxiapp.ui.listvehicles

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.mytaxiapp.R
import com.mytaxiapp.data.model.Placemark
import com.mytaxiapp.ui.map.MapActivity
import com.mytaxiapp.util.Logger

import java.util.ArrayList


/**
 * @author Tosin Onikute.
 */

class VehicleListAdapter(private val context: Context, private val placemarks: ArrayList<Placemark>?) : RecyclerView.Adapter<VehicleListAdapter.ViewHolder>() {

    private val logger = Logger.getLogger(javaClass)
    private val mTypedValue = TypedValue()
    private val mBackground: Int

    private var caddress: String? = null
    private var cname: String? = null
    private var cinterior: String? = null
    private var cfuel: String? = null

    private val INTERIOR = "INTERIOR"
    private val VIN = "VIN"
    private val FUEL = "FUEL"
    private val ENGINE = "ENGINE"

    inner class ViewHolder(val mView: View) : RecyclerView.ViewHolder(mView) {
        val textAddress: TextView
        val textName: TextView
        val textInterior: TextView
        val textFuel: TextView

        init {

            textAddress = mView.findViewById(R.id.address) as TextView
            textName = mView.findViewById(R.id.name) as TextView
            textInterior = mView.findViewById(R.id.interior_vin) as TextView
            textFuel = mView.findViewById(R.id.fuel_engine) as TextView
        }

        override fun toString(): String {
            return super.toString()
        }
    }

    fun getValueAt(position: Int): String {
        return placemarks!![position].name.toString()
    }

    init {
        context.theme.resolveAttribute(R.attr.selectableItemBackground, mTypedValue, true)
        mBackground = mTypedValue.resourceId
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.vehicle_list, parent, false)
        view.setBackgroundResource(mBackground)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        /* Set your values */
        val model = placemarks!![position]

        caddress = ""
        cname = ""
        cinterior = ""
        cfuel = ""

        if (model.address != null) {
            caddress = model.address
        }
        if (model.name != null) {
            cname = model.name
        }
        if (model.interior != null && model.vin != null) {
            cinterior = INTERIOR + " : " + model.interior + " | " + VIN + " : " + model.vin
        }
        if (model.fuel != null && model.engineType != null) {
            cfuel = FUEL + " : " + model.fuel + "  |  " + ENGINE + " : " + model.engineType
        }

        holder.textAddress.text = caddress
        holder.textName.text = cname
        holder.textInterior.text = cinterior
        holder.textFuel.text = cfuel

        // launch the detail activity to show vehicle information
        holder.mView.setOnClickListener { v ->
            val context = v.context
            val intent = Intent(context, MapActivity::class.java)
            intent.putExtra("position", holder.adapterPosition)
            val activity = v.context as Activity
            activity.startActivityForResult(intent, 500)
            activity.overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left)
        }

    }

    override fun getItemCount(): Int {
        return placemarks?.size ?: 0
    }

    fun addAll(data: List<Placemark>) {
        notifyDataSetChanged()
    }

    fun add(data: Placemark) {
        notifyDataSetChanged()
        placemarks!!.add(data)
    }


    fun getItemPos(pos: Int): Placemark {
        return placemarks!![pos]
    }

    fun clear() {
        placemarks!!.clear()
    }

}

